app/scripts/modules/core/account/account.service.js

```diff
@@ -11,38 +11,37 @@ module.exports = angular.module('spinnaker.core.account.service', [
 ])
   .factory('accountService', function(settings, _, Restangular, $q, infrastructureCaches, cloudProviderRegistry) {
 
-    function getPreferredZonesByAccount(providerName='aws') {
-      return $q.when(settings.providers[providerName].preferredZonesByAccount);
-    }
+    let getAllAccountDetailsForProvider = _.memoize((providerName) => {
+      return listAccounts(providerName)
+        .then((accounts) => $q.all(accounts.map((account) => getAccountDetails(account.name))));
+    });
 
-    function getAvailabilityZonesForAccountAndRegion(providerName, accountName, regionName) {
-      return getPreferredZonesByAccount(providerName).then( function(defaults) {
-        if (defaults[accountName] && defaults[accountName][regionName]) {
-          return {preferredZones: defaults[accountName][regionName]};
-        }
-        if (!defaults[accountName] && defaults.default && defaults.default[regionName]) {
-          return {preferredZones: defaults.default[regionName]};
-        }
-        return {preferredZones: []};
-      })
-      .then(function(zonesCollection) {
-        return getRegionsForAccount(accountName).then(function(regions){
-          //TODO(chrisberry)
-          //Make this pluggable to remove the need for provider tests
-          if (providerName === 'azure') {
-            zonesCollection.actualZones = [regionName];
-          }
-          else {
-            zonesCollection.actualZones = _.find(regions, {name: regionName}).availabilityZones;
-          }
-          return zonesCollection;
+    let getPreferredZonesByAccount = _.memoize((providerName) => {
+      let preferredZonesByAccount = {};
+      return getAllAccountDetailsForProvider(providerName)
+        .then((accounts) => {
+          accounts.forEach((account) => {
+            preferredZonesByAccount[account.name] = {};
+            account.regions.forEach((region) => {
+              let preferredZones = region.availabilityZones;
+              //TODO(chrisberry)
+              //Make this pluggable to remove the need for provider tests
+              if (providerName === 'azure') {
+                preferredZones = [region.name];
+              }
+              if (region.preferredZones) {
+                preferredZones = _.intersection(region.preferredZones, region.availabilityZones);
+              }
+              preferredZonesByAccount[account.name][region.name] = preferredZones;
+            });
+          });
+          return preferredZonesByAccount;
```

Comment:
Usually like to see a 'catch' block on promises. Either here or on the "getAllAccountDetailsForProvider" call.